test/filewriter.spec.ts

```diff
@@ -1,41 +1,88 @@
 import fs from "fs";
 import { openSlpFile, SlpInputSource } from "../src/utils/slpReader";
-import { SlpFileWriter } from "../src";
+import { SlpFileWriter, SlippiGame } from "../src";
 import { Writable } from "stream";
 
+// On my machine, >100 is required to give the slpFile.ts "finish" callback time to execute.
+// I thought a 'yield' 0 ms setTimout would allow the callback to execute, but that's not the case.
+const TIMEOUT_MS = 1000;
+
 describe("when ending SlpFileWriter", () => {
   it("should write data length to file", async () => {
-    const testFilePath = "slp/finalizedFrame.slp";
-
-    const slpFileWriter = new SlpFileWriter();
-    const slpFile = openSlpFile({ source: SlpInputSource.FILE, filePath: testFilePath });
-    const dataLength = slpFile.rawDataLength;
-    const dataPos = slpFile.rawDataPosition;
+    const { dataLength, newFilename } = runSlpFileWriter("slp/finalizedFrame.slp");
+    await new Promise((resolve) => {
+      setTimeout(() => {
+        const writtenDataLength = openSlpFile({ source: SlpInputSource.FILE, filePath: newFilename }).rawDataLength;
+        fs.unlinkSync(newFilename);
 
-    const testFd = fs.openSync(testFilePath, "r");
-    const newPos = pipeMessageSizes(testFd, dataPos, slpFileWriter);
+        expect(writtenDataLength).toBe(dataLength);
 
-    const newFilename = slpFileWriter.getCurrentFilename();
-    const buffer = Buffer.alloc(4);
+        resolve();
+      }, TIMEOUT_MS);
+    });
+  });
 
-    pipeAllEvents(testFd, newPos, dataPos + dataLength, slpFileWriter, slpFile.messageSizes);
+  it("should succeed if no display names or connect codes are available in game start", async () => {
+    const { newFilename } = runSlpFileWriter("slp/finalizedFrame.slp");
     await new Promise((resolve) => {
-      // On my machine, >100 is required to give the slpFile.ts "finish" callback time to execute.
-      // I thought a 'yield' 0 ms setTimout would allow the callback to execute, but that's not the case.
-      const timeoutMs = 1000;
+      setTimeout(() => {
+        const players = new SlippiGame(newFilename).getMetadata().players;
+        Object.keys(players).forEach((key) => {
+          const player = players[key];
+          players[key] = player.names;
+        });
+        fs.unlinkSync(newFilename);
+
+        expect(players).toEqual({ 0: { netplay: "", code: "" }, 1: { netplay: "", code: "" } });
```

Comment:
I'm not sure if this is good practice or personal preference, but I think if we know something does not exist, it should be null, rather than an empty string. Like `player.names` should return null for replays from console/vs mode. Thoughts on having `null` as opposed to the `{ netplay: string, code: string }` object?

Reply:
my understanding of best practices is that we should avoid `null` and nullable fields/vars wherever possible. I assumed that turning on strict null checks in https://github.com/project-slippi/slippi-js/commit/f3c4ca997c518c8924f45fe2176c0ac622471ceb expressed a preference toward removing `null` from the codebase completely in the long run

Reply:
That's sound enough logic for keeping them not nullable.

A different concern I had was there currently exists a `player.nametag` field which contains their in-game name tag if applicable (in offline/VS mode). Should these fields not be combined together? Whether that looks like or something else:
```
names: {
  netplay: "",
  tag: "",
  code: "",
}
```
I'm still not entirely sold on any of these names maybe aside from "code". Since in-game it's called `Name entry` so maybe the in-game name should be called `name`, but then having `player.names.name` field seems kinda really dumb.

Reply:
you mean in `types.ts`?

what purpose would that serve?

Reply:
I think my gripe was with `names` being the top level key but then the in-game tag is also a name but isn't in the object, and technically `code` isn't really a name. But if we just rename the top level key to `netplay` and the in netplay display name field to `name` I'd be fine with it.